File: T1/ex-2-vec-vectorized.f90
! Optimization Example
! v1.0 - 2012/02/01 - carlos@tacc.utexas.edu

PROGRAM main
  IMPLICIT NONE
  INTEGER, PARAMETER :: MAX_SIZE = 50000
  INTEGER :: i, k, rate, tStart, tSetup, tStop
  REAL    :: setup, kernel
  REAL    :: x(1:MAX_SIZE)
  REAL    :: y(1:MAX_SIZE)
  

  setup  = 0.0
  kernel = 0.0
  CALL SYSTEM_CLOCK( COUNT_RATE = rate )
  ! External loop is only used to make things take longer
  ! DO NOT CHANGE THIS EXTERNAL LOOP
  DO k = 1, 1000

    ! Initialize the arrays
    CALL SYSTEM_CLOCK( COUNT = tStart )
    DO i = 1, MAX_SIZE
        x(i) = 2.0*i
    END DO
    CALL SYSTEM_CLOCK( COUNT = tSetup )

    DO i = 1, MAX_SIZE
        y(i) = x(i)
    END DO

    ! KERNEL
    DO i = 2, MAX_SIZE-1
        x(i) = y(i-1)/4 + y(i)/2 + y(i+1)/4;
    END DO

    CALL SYSTEM_CLOCK( COUNT = tStop )

  setup  = setup  + REAL( ( tSetup - tStart ) ) / REAL( rate )
  kernel = kernel + REAL( ( tStop - tSetup ) ) / REAL( rate )
  END DO

  WRITE(*,*)'Two values for validation'
  WRITE(*,*)'========================='
  WRITE(*,*)'X(2) = ',x(2)
  WRITE(*,*)'X(',MAX_SIZE-1,') = ',x(MAX_SIZE-1)
  WRITE(*,*)
  WRITE(*,*)'Execution Times'
  WRITE(*,*)'==============='
  WRITE(*,*)'Setup Time  = ',setup
  WRITE(*,*)'Kernel Time = ',kernel
  WRITE(*,*)

 END PROGRAM

File: T1/ex-2-vec-vectorized.c
// Optimization Example
// v1.0 - 2010/03/20 - carlos@tacc.utexas.edu
// 
// Example of a naively written code
// This code contains data dependencies that prevent vecorization, 
// as well as type-mixed math and a naive mathematical implementation.
// Optimize the computational kernel in lines 34-36 of this file
// as much as possible

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define MAX_SIZE 50000

int main(int argc, char *argv[])
{
  int i, j, k;
  double tStart, tSetup, tEnd, setup, kernel;
  double x[MAX_SIZE];
  double y[MAX_SIZE];

  setup = 0.0;
  kernel = 0.0;
  /* External loop is only used to make things take longer */
  /* DO NOT CHANGE THIS EXTERNAL LOOP */
  for( k = 1; k < 1000; k++){

    /* Initialize the arrays */
    tStart = (double)clock()/(double)CLOCKS_PER_SEC;
    for(i = 0; i < MAX_SIZE; i++){
        x[i] = 2.0*i;
    }
    tSetup = (double)clock()/(double)CLOCKS_PER_SEC;

    /* KERNEL */
    for(i = 0; i < MAX_SIZE; i++){
	y[i] = x[i];
    }

    for(i = 1; i < MAX_SIZE-1; i++){
        x[i] = y[i-1]/4 + y[i]/2 + y[i+1]/4;
    }
    tEnd = (double)clock()/(double)CLOCKS_PER_SEC;

  setup  += tSetup - tStart;
  kernel += tEnd - tSetup;
  }

  printf( "\nTwo values for validation");
  printf( "\n=========================");
  printf( "\nX[1]  = %e", x[1]);
  printf( "\nX[%d] = %e\n", MAX_SIZE-2, x[MAX_SIZE-2]);

  printf( "\nExecution times");
  printf( "\n=========================");
  printf( "\nSetup Time  = %le", setup );
  printf( "\nKernel Time = %le\n\n", kernel );



return 0;
}
